Add kvstore based node discovery

tgraf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4270)
<!-- Reviewable:end -->

aanm commented on Documentation/kvstore.rst:
```diff
@@ -0,0 +1,70 @@
+Prefix                Description
+===================== ====================
+``cilium/``           All keys share this common prefix.
+``cilium/state/``     State stored by agents, data is automatically recreated on removal or corruption.
```
Maybe it is better to specify this data is going to be recreated per agent. At a first sight it looked it all prefix was going to be recreated on removal / corruption

tgraf replied:
Given we are not using `/conf/` yet, I've removed this from the doc.

aanm commented on Documentation/kvstore.rst:
```diff
@@ -0,0 +1,70 @@
+``cilium/state/``     State stored by agents, data is automatically recreated on removal or corruption.
+``cilium/conf/``      Configuration, agents will read this to retrieve
+                      configuration but will never write to these keys.
+                      Used to store agent configuration in the kvstore.
```
For me, this sentence does not seems right. With the previous sentence you say all "agents will read" but here you say "used to store agent configuration".

If the agents, AFAIK, are the only client to the kvstore, who's going to store that agent configuration?

tgraf replied:
Given we are not using `/conf/` yet, I've removed this from the doc.

"stored" does not mean that that the agent writes to it. It could be another tool leaving behind configuration for all agents.

aanm commented on daemon/k8s_watcher.go:
```diff
@@ -443,46 +441,6 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
 
 	si.Start(wait.NeverStop)
 
-	_, nodesController := cache.NewInformer(
```
What happen here? Why are we no longer watching for node events? How are we going to retrieve the podCIDR?

tgraf replied:
The PodCIDR is retrieved independent of the node watcher, here:
https://github.com/cilium/cilium/blob/master/pkg/k8s/init.go#L112

aanm commented on pkg/node/local_node.go:
```diff
@@ -0,0 +1,50 @@
+		IPAddresses: []Address{
+			{
+				AddressType: v1.NodeInternalIP,
+				IP:          GetExternalIPv4(),
```
`GetExternalIPv4` can be nil on a v6 only node?

tgraf replied:
This is existing code that is only moved but I think it is never nil:

```
func InitDefaultPrefix(device string) {
[...]

        ip, err := firstGlobalV4Addr(device)
        if err != nil {
                return
        }

        if ipv4ExternalAddress == nil {
                ipv4ExternalAddress = ip
        }
```

aanm commented on Documentation/cmdref/cilium.md:
```diff
@@ -27,6 +27,7 @@ CLI for interacting with the local Cilium Agent
 * [cilium identity](cilium_identity.html)	 - Manage security identities
 * [cilium kvstore](cilium_kvstore.html)	 - Direct access to the kvstore
 * [cilium monitor](cilium_monitor.html)	 - Display BPF program events
+* [cilium node](cilium_node.html)	 - Manage cluster nodes
```
I like the commit message with the ":100:"

tgraf committed:
node: Provide `cilium node list` command

```
$ cilium node list
Name                                          IPv4 Address    Endpoint CIDR   IPv6 Address   Endpoint CIDR
ip-172-0-117-60.us-west-2.compute.internal    172.0.117.60    10.2.2.0/24     <nil>          f00d::a02:200:0:0/112
ip-172-0-186-231.us-west-2.compute.internal   172.0.186.231   10.2.3.0/24     <nil>          f00d::a02:300:0:0/112
ip-172-0-50-227.us-west-2.compute.internal    172.0.50.227    10.2.0.0/24     <nil>          f00d::a02:0:0:0/112
ip-172-0-51-175.us-west-2.compute.internal    172.0.51.175    10.2.1.0/24     <nil>          f00d::a02:100:0:0/112
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Add kvstore based node discovery

Each agent stores its own node information in the kvstore by using a shared
store. This enables agents to discover others nodes by joining a shared store.

Individual nodes belong to a cluster with a fixed name for name. Future
versions will make this configurable.

Signed-off-by: Thomas Graf <thomas@cilium.io>
